libs/ledger-live-common/src/families/cosmos/js-synchronisation.ts

```diff
@@ -54,130 +54,110 @@ const txToOps = (info: AccountShapeInfo, accountId: string, txs: CosmosTx[]): Op
+          const validator = message.attributes.find(attr => attr.key === "validator")?.value;
+          if (amount && validator && amount.indexOf(currency.units[1].code) != -1) {
+            delegateShards.push({
+              amount: new BigNumber(amount.replace(currency.units[1].code, "")),
+              address: validator,
+            });
+          }
         }
+        op.extra.validators = delegateShards;
         break;
-
-      case "redelegate":
-        if (
-          attributes.amount &&
-          attributes.amount.indexOf(currency.units[1].code) != -1 &&
-          attributes.destination_validator &&
-          attributes.source_validator
-        ) {
-          op.type = "REDELEGATE";
-          op.value = new BigNumber(fees);
-          op.extra.validators.push({
-            address: attributes.destination_validator,
-            amount: attributes.amount.replace(currency.units[1].code, ""),
-          });
-          op.extra.sourceValidator = attributes.source_validator;
+      }
+      case "redelegate": {
+        op.type = "REDELEGATE";
+        op.value = new BigNumber(fees);
+        const redelegateShards: { amount: BigNumber; address: string }[] = [];
+        for (const message of correspondingMessages) {
+          const amount = message.attributes.find(attr => attr.key === "amount")?.value;
+          const validatorDst = message.attributes.find(attr => attr.key === "validator_dst")?.value;
+          const validatorSrc = message.attributes.find(attr => attr.key === "validator_src")?.value;
+          if (
+            amount &&
+            validatorDst &&
+            validatorSrc &&
+            amount.indexOf(currency.units[1].code) != -1
+          ) {
+            op.extra.sourceValidator = validatorSrc;
+            redelegateShards.push({
+              amount: new BigNumber(amount.replace(currency.units[1].code, "")),
+              address: validatorDst,
+            });
+          }
         }
+        op.extra.validators = redelegateShards;
         break;
-
-      case "unbond":
-        if (
-          attributes.amount &&
-          attributes.amount.indexOf(currency.units[1].code) != -1 &&
-          attributes.validator
-        ) {
-          op.type = "UNDELEGATE";
-          op.value = new BigNumber(fees);
-          op.extra.validators.push({
-            address: attributes.validator,
-            amount: attributes.amount.replace(currency.units[1].code, ""),
-          });
+      }
+      case "unbond": {
+        op.type = "UNDELEGATE";
+        op.value = new BigNumber(fees);
+        const unbondShards: { amount: BigNumber; address: string }[] = [];
+        for (const message of correspondingMessages) {
+          const amount = message.attributes.find(attr => attr.key === "amount")?.value;
+          const validator = message.attributes.find(attr => attr.key === "validator")?.value;
+          if (amount && validator && amount.indexOf(currency.units[1].code) != -1) {
```

Comment:
We could add a variable named `const validAmountUnit = amount.indexOf(currency.units[1].code) !== -1`

Minor suggestion

Same in other op type

Reply:
Thanks a lot
I've done some minor refactoring to make the code more concise